Added Ceph Dashboard Changes

sunilangadi2 opened the pull request:
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs

# Checklist:

- [x] Create a test case in Polarion reviewed and approved.
- [x] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [x] Review the automation design
- [x] Implement the test script and perform test runs
- [x] Submit PR for code review and approve
- [x] Update Polarin Test with Automation script details and update automation fields
- [x] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test

Hi Guys,
Added ceph dashboard changes, Please Review it.

**CLI used**:  `python run.py --rhbuild 4.1 --global-conf conf/nautilus/dashboard/sanity-dashboard.yaml --osp-cred osp/osp-cred-ci-2.yaml --inventory conf/inventory/rhel-8.2-server-x86_64.yaml --suite suites/nautilus/dashboard/sanity_containerized_ceph_dashboard.yaml --log-level info   --insecure-registry --docker-registry registry.redhat.io --docker-image rhceph/rhceph-4-rhel8 --docker-tag latest --ignore-latest-container --post-results
`

User can access the ceph dashboard with _Active mgr node ip_ with 8443 port 
e.g : http://10.0.103.19:8443/#/dashboard
Success logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1596788796485/
Polarion Link : https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83573534

**NOTE**: These tests requires docker credentials and user has to set dashboard and grafana credentials here
suites/nautilus/dashboard/sanity_containerized_ceph_dashboard.yaml

![dashboard_purge](https://user-images.githubusercontent.com/22169979/89633881-15d8b480-d8c2-11ea-91eb-bb68a25ef606.png)

>

rakeshgm commented:
@sunilangadi2 good one, this is just the installation i believe, can you add this in sanity ansible suite unless there is specific test.

sunilangadi2 commented:
@sunilkumarn417 Added purge test.

sunilangadi2 commented:
> @sunilangadi2 good one, this is just the installation i believe, can you add this in sanity ansible suite unless there is specific test.
> @rakeshgm This require some prereq like credentials so it's good to keep it in separate folder otherwise ansible jenkins get failed

XDsubhash commented on suites/nautilus/dashboard/sanity_containerized_ceph_dashboard.yaml:
```diff
@@ -0,0 +1,104 @@
+          is_mixed_lvm_configs: True
+          ansi_config:
+              ceph_test: True
+              ceph_origin: distro
+              ceph_stable_release: nautilus
+              ceph_repository: rhcs
+              osd_scenario: lvm
+              osd_auto_discovery: False
+              journal_size: 1024
+              ceph_stable: True
+              ceph_stable_rh_storage: True
+              containerized_deployment: true
+              copy_admin_key: true
+              ceph_docker_registry: "registry.redhat.io"
+              ceph_docker_registry_auth: true
+              ceph_docker_registry_username: <provide_username>
+              ceph_docker_registry_password: <provide_password>
+              dashboard_enabled: True
+              dashboard_admin_user: <provide_username>
+              dashboard_admin_password: <provide_password>
+              grafana_admin_user: <provide_username>
+              grafana_admin_password: <provide_password>
+              node_exporter_container_image: registry.redhat.io/openshift4/ose-prometheus-node-exporter:v4.1
+              grafana_container_image: registry.redhat.io/rhceph/rhceph-4-dashboard-rhel8:4
+              prometheus_container_image: registry.redhat.io/openshift4/ose-prometheus:4.1
+              alertmanager_container_image: registry.redhat.io/openshift4/ose-prometheus-alertmanager:4.1
```
ceph_origin: distro,  unreleased rhceph version

with 
grafana_container_image: registry.redhat.io/rhceph/rhceph-4-dashboard-rhel8:4
released dashboard version

is it intented @sunilangadi2 ?

sunilangadi2 replied:
@XDsubhash modified ceph origin to **repository** with the same released grafana image

XDsubhash replied:
@sunilangadi2 is this suite supposed to test GAed versions? or internal builds?

sunilangadi2 replied:
@XDsubhash this dashboard suite requires `registry registry.redhat.io` and suppose if i gave this registry  `registry-proxy.engineering.redhat.com` then dashboard installation fails so currently this suite works fine with GAed version once we figure out how to do automation of podman login of this registry registry.redhat.io to all nodes https://github.com/red-hat-storage/cephci/issues/189 then this suite also works for distro (means internal builds)

sunilkumarn417 replied:
@sunilangadi2 #189 issue is to handle the same registry(registry.redhat.io) that  you are addressing in previous comment.

sunilangadi2 replied:
@sunilkumarn417 yes if we resolve this issue https://github.com/red-hat-storage/cephci/issues/189 then we can pass this registry `registry-proxy.engineering.redhat.com` in cli

rakeshgm replied:
issue linked.

XDsubhash commented on suites/nautilus/dashboard/sanity_containerized_ceph_dashboard.yaml:
```diff
@@ -0,0 +1,104 @@
+              ceph_repository: rhcs
+              osd_scenario: lvm
+              osd_auto_discovery: False
+              journal_size: 1024
```
remove journal_size if its not needed in lvm bluestore osd cases

sunilangadi2 replied:
@XDsubhash yes removed journal_size as it's not needed

sunilkumarn417 commented:
@XDsubhash , @rakeshgm . added a git issue for handling creds of registry.redhat.io registry.
https://github.com/red-hat-storage/cephci/issues/189

sunilangadi2 commented on suites/nautilus/dashboard/sanity_containerized_ceph_dashboard.yaml:
```diff
@@ -0,0 +1,103 @@
+              ceph_docker_registry_username: <provide_username>
+              ceph_docker_registry_password: <provide_password>
+              dashboard_enabled: True
+              dashboard_admin_user: admin
+              dashboard_admin_password: p@ssw0rd
+              grafana_admin_user: admin
+              grafana_admin_password: p@ssw0rd
```
Added default username and password so that once https://github.com/red-hat-storage/cephci/issues/189 this issue got fixed we can also add this to Jenkins,Please let me know Is it good to merge now?

rakeshgm replied:
Sunil Angadi, Can you take up this issue #189

sunilangadi2 replied:
@rakeshgm @XDsubhash I think subhash has some workaround, Sure I will take it up.

rakeshgm replied:
OK. Csn you assign the issue to yourself.

sunilangadi2 replied:
yes assigned on my name

rakeshgm commented:
@sunilangadi2 are you able finish the issue #189

sunilangadi2 commented:
@rakeshgm It is In-progress, I will send the PR soon

sunilangadi2 commented:
dashboard suite is good to go with Jenkins.!!

XDsubhash commented on tests/misc_env/install_prereq.py:
```diff
@@ -130,3 +132,22 @@ def enable_rhel_rpms(ceph, distro_ver):
+    user = cdn_cred.get('username')
+    pwd = cdn_cred.get('password')
+
+    conatiner = 'docker'
```
update spelling

sunilangadi2 replied:
@XDsubhash Sorry, I will do that.

XDsubhash commented on tests/misc_env/install_prereq.py:
```diff
@@ -65,6 +66,7 @@ def install_prereq(ceph, timeout=1800, skip_subscription=False, repo=False, rhbu
             ceph.exec_command(cmd='sudo pip install crefi', long_running=True)
         ceph.exec_command(cmd='sudo yum clean metadata')
         config_ntp(ceph)
+    registry_login(ceph, distro_ver)
```
did you check negative cases ? if other suites fail or pass if the cdn_credentials vars are not provided or updated  cephci.yaml.template doesnt exists?

sunilangadi2 replied:
@XDsubhash Yes I have tested negative cases, if cdn_credentials not provided in cephci.yaml file then install_prereq will fail, I think It's good to have this as compulsory task in prereq so that other suites can perform cdn container testing with this logged in registry also this logged in registry doesn't affect with distro registry i.e registry-proxy.engineering.redhat.com
we just have to update everyone to provide these credentials before running and they can trigger any distro or any container cdn suite

rakeshgm replied:
the failure should not come for every execution. the failure should be seen only if its a container suite . why would we this if we are not running non-container tests without dashboard

sunilangadi2 replied:
@rakeshgm for dashboard both bare-metal and container, we should be doing login to 'registry.redhat.io' so i made it as compulsory task in prereq since this is not required for non-container tests without dashboard, I will add some checks not to fail for rpm based tests without cdn credentials, Thanks for pointing it out

XDsubhash commented on tests/misc_env/install_prereq.py:
```diff
@@ -130,3 +132,22 @@ def enable_rhel_rpms(ceph, distro_ver):
+
+    ceph.exec_command(cmd='sudo yum install -y {c}'.format(c=conatiner), long_running=True)
+
+    ceph.exec_command(cmd='sudo systemctl restart {c}'.format(c=conatiner), long_running=True)
```
why are we restarting here?  
does podman have daemon?

sunilangadi2 replied:
@XDsubhash sorry, I will make this restart only to docker, because I noticed some time docker service works only after restarting.

XDsubhash commented on tests/misc_env/install_prereq.py:
```diff
@@ -130,3 +132,22 @@ def enable_rhel_rpms(ceph, distro_ver):
+    ceph.exec_command(cmd='sudo systemctl restart {c}'.format(c=conatiner), long_running=True)
+
+    ceph.exec_command(
+        cmd='sudo {c} login -u {u} -p {p} registry.redhat.io'.format(c=conatiner, u=user, p=pwd), long_running=True)
```
after reboot TC's ..do we have to log in again ?

sunilangadi2 replied:
@XDsubhash After node reboot no need to log in again.

XDsubhash replied:
okay

XDsubhash commented on cephci.yaml.template:
```diff
@@ -8,6 +8,10 @@ polarion:
   username: some_user
   password: some_password
 
+cdn_credentials:
```
can you add couple comments saying for which sort of suites its a hard requirement ?

eg:
Provide these credentials for suites where we need to access cdn content like GAed container image registry log in etc..

sunilangadi2 replied:
@XDsubhash Thanks for the review sure I will address this.

sunilangadi2 commented:
Addressed Comments
Now tests won't fail if cdn_credentials not provided
tests with cdn_credentials provided: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1599128411673/result.html
tests without cdn_credentials provided: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1599118518926/result.html

rakeshgm commented on tests/misc_env/install_prereq.py:
```diff
@@ -130,3 +132,26 @@ def enable_rhel_rpms(ceph, distro_ver):
+        if not (user and pwd):
+            return
+
+        container = 'docker'
+        if distro_ver.startswith('8'):
+            container = 'podman'
+
+        ceph.exec_command(cmd='sudo yum install -y {c}'.format(c=container), long_running=True)
+
+        if container == 'docker':
+            ceph.exec_command(cmd='sudo systemctl restart docker', long_running=True)
+
+        ceph.exec_command(
+            cmd='sudo {c} login -u {u} -p {p} registry.redhat.io'.format(c=container, u=user, p=pwd), long_running=True)
```
These look still redundant code for me. lot of `'if`s are present. you could try to use a dictionary here.

sunilangadi2 replied:
@rakeshgm yes I agree with using dictionary where large items are there to check using 'if' but here two items also even using dictionary needs two checks also usage of dictionary makes slow down the performance of the code with so many opcode lookup operation, I feel it's ok with two 'if',  Please let me know your thoughts?

rakeshgm commented on tests/misc_env/install_prereq.py:
```diff
@@ -130,3 +132,26 @@ def enable_rhel_rpms(ceph, distro_ver):
+        user = cdn_cred.get('username')
+        pwd = cdn_cred.get('password')
+        if not (user and pwd):
+            return
```
if you have credentials are missing and when they are needed for the scenario, how are you throwing error. you are just returning without any message.

sunilangadi2 replied:
@rakeshgm sure I will handle that error

sunilangadi2 commented:
Handled exception if user didn't specified cdn credentials It shows warning in the logs

http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1599626576389/result.html
`install_prereq - WARNING - cdn_credentials not found`
also the successful log
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1599628498233/result.html

sunilkumarn417 commented:
@sunilangadi2 , can you add more details of this PR and updates in the Description, since this code changes consists #189 

@rakeshgm , @XDsubhash code changes LGTM.

sunilangadi2 commented:
@sunilkumarn417 Thanks, Updated the PR description.
@rakeshgm and @XDsubhash Please let me know your views on it?

rakeshgm reviewed (commented):
looks good,

rakeshgm commented:
@sunilangadi2  squash to meaningful comments.
